dotnet/src/SemanticKernel/Planning/SKContextExtensions.cs

```diff
@@ -2,47 +2,162 @@
-        List<string>? excludedSkills = null,
-        List<string>? excludedFunctions = null)
+        PlannerSkillConfig config,
+        string? semanticQuery = null)
     {
-        excludedSkills ??= new();
-        excludedFunctions ??= new();
+        var excludedSkills = config.ExcludedSkills ?? new();
+        var excludedFunctions = config.ExcludedFunctions ?? new();
+        var includedFunctions = config.IncludedFunctions ?? new();
 
         context.ThrowIfSkillCollectionNotSet();
 
         var functionsView = context.Skills!.GetFunctionsView();
 
-        return functionsView.SemanticFunctions
+        var availableFunctions = functionsView.SemanticFunctions
             .Concat(functionsView.NativeFunctions)
             .SelectMany(x => x.Value)
             .Where(s => !excludedSkills.Contains(s.SkillName) && !excludedFunctions.Contains(s.Name))
             .ToList();
+
+        List<FunctionView> result = availableFunctions;
+        if (!string.IsNullOrEmpty(semanticQuery))
+        {
+            // If a Memory provider has not been registered, do not filter the functions.
+            if (context.Memory is not NullMemory)
+            {
+                // catalog functions to memory
+                foreach (var function in availableFunctions)
+                {
+                    var functionName = function.ToFunctionName();
+                    var key = string.IsNullOrEmpty(function.Description) ? functionName : function.Description;
+
+                    // It'd be nice if there were a saveIfNotExists method on the memory interface
+                    var memoryEntry = await context.Memory.GetAsync(MemoryCollectionName, key, context.CancellationToken);
+                    if (memoryEntry == null)
+                    {
+                        // TODO It'd be nice if the minRelevanceScore could be a parameter for each item that was saved to memory
+                        // As folks may want to tune their functions to be more or less relevant.
+                        await context.Memory.SaveInformationAsync(MemoryCollectionName, key, functionName, function.ToManualString(),
+                            context.CancellationToken);
+                    }
+                }
+
+                var memories = context.Memory.SearchAsync(MemoryCollectionName, semanticQuery, config.MaxFunctions, config.RelevancyThreshold,
+                    context.CancellationToken);
+
+                result = new List<FunctionView>();
+                await foreach (var memoryEntry in memories)
+                {
+                    var function = availableFunctions.Find(x => x.ToFunctionName() == memoryEntry.Id);
+                    if (function != null)
+                    {
+                        context.Log.LogDebug("Found relevant function. Relevance Score: {0}, Function: {1}", memoryEntry.Relevance, function.ToFunctionName());
+                        result.Add(function);
+                    }
+                }
+
```

Comment:
Lots of nest looping going on here (Find and Any in foreach)... having a hard time following. Does it make sense to refactor these out into private helpers, or at least add a comment in each section explaining what it's doing next.

Reply:
Did some refactoring here.